### PR TITLE
Some general Vector3 optimizations

### PR DESCRIPTION
**Pull request by spydon**
Since the `Vector3.zero` constructor was already a normal named constructor it was already possible to extend the class.

 - ~~Converts all factory constructors (except `Vector3.random`) to named constructors that directly sets the storage instead of calling intermediary methods to do it.~~
 - Simplifies `length2`, `dot`, `isInfinite` and `isNaN`.
 - Optimizes `reflect`, `relativeError` and `absoluteError` so that they don't create extra `Vector3` objects.
 - Orders all operations made on `_v3storage` from 1 to 0 so that it is consistent across the class and guarantees that bounds checking is only done once.

### CODE REVIEW

**Comment by spydon** [comment]
@johnmccutchan fixed up `Vector3` here, I'll hopefully have time to fix `Vector4` tomorrow.

File: lib/src/vector_math/vector3.dart [hunk 1]
```diff
@@ -555,46 +555,46 @@ class Vector3 implements Vector {
...
-    _v3storage[0] = argStorage[0];
+    final argStorage = arg.storage;
     _v3storage[1] = argStorage[1];
+    _v3storage[0] = argStorage[0];
   }
 
   set xz(Vector2 arg) {
-    final argStorage = arg._v2storage;
-    _v3storage[0] = argStorage[0];
+    final argStorage = arg.storage;
```
Comment (johnmccutchan) [inline lib/src/vector_math/vector3.dart:564]: please restore the ._v2storage accessor here

File: lib/src/vector_math/vector3.dart [hunk 2]
```diff
@@ -555,46 +555,46 @@ class Vector3 implements Vector {
...
     _v3storage[2] = argStorage[1];
+    _v3storage[0] = argStorage[0];
   }
 
   set yx(Vector2 arg) {
-    final argStorage = arg._v2storage;
+    final argStorage = arg.storage;
```
Comment (johnmccutchan) [inline lib/src/vector_math/vector3.dart:570]: here and below

File: lib/src/vector_math/vector3.dart (hunk 1 above)
Comment (spydon) [reply lib/src/vector_math/vector3.dart]: Ah, was sure I had set all of them back, found some for the matrix too.
They should all be fixed now!

Comment (spydon) [reply lib/src/vector_math/vector3.dart]: Oops, shouldn't have done that in vim, one more fix coming up!

Comment (spydon) [reply lib/src/vector_math/vector3.dart]: Now it should be fine

File: lib/src/vector_math/vector3.dart [hunk 3]
```diff
@@ -332,200 +338,194 @@ class Vector3 implements Vector {
...
 
   /// Create a copy of this and scale it by [arg].
   Vector3 scaled(double arg) => clone()..scale(arg);
 
   /// Negate this.
   void negate() {
-    _v3storage[2] = -_v3storage[2];
-    _v3storage[1] = -_v3storage[1];
-    _v3storage[0] = -_v3storage[0];
+    _v3storage[2] *= -1;
```
[the remark this answers is not in the document]
Comment (spydon) [reply lib/src/vector_math/vector3.dart:466]: Alright, then I have to do that for the already merged Vector2-PR (#289) too.

[the remark this answers is not in the document]
Comment (spydon) [reply lib/src/vector_math/vector3.dart:466]: Fixed Vector2 in #296

**Comment by rakudrama** [comment]
> Since the `Vector3.zero` constructor was already a normal named constructor it was already possible to extend the class.

I think we will want to seal this class to prevent it being extended. This will make performance and behaviour more reliable.

The classes in vector_math _were never designed to be extended_.  There are no guarantees which methods are implemented in terms of other methods to provide an extension point. When `Vector3` is extended, most of the behaviour is working by accident rather than by design. If `Vector3` was designed to be extended, this would be documented. Say it was the intention to allow a subclass to implement some kind of observable `Vector3`. Then _all_ mutating methods would have been implemented in terms of `setValue` so that `setValue` could be overridden as a single place to detect changes.

**Dart 3** now has [class modifiers](https://dart.dev/language/class-modifiers) to prevent classes being extended or implemented. When vector_math is revised to be a Dart 3 library, the ability to extend `Vector3` should be removed.

The efficiency concern is that extending a class makes calls to the base class methods be polymorphic. When the class is not extended, calls to a method can be resolved exactly and inlined or implemented as direct calls. When the methods  have overrides, the calls are necessarily indirect calls as there may be more than one version of the method. For a simple class like `Vector3` this can be a significant performance penalty. Say there is an app that uses `Vector3` and is fast. One day the app is changed to import a library that extends `Vector3`. All the existing code becomes slower because of the calls that become indirect (and can no longer be optimized).

> 
> * Converts all factory constructors (except `Vector3.random`) to named constructors that directly sets the storage instead of calling intermediary methods to do it.

I worry that these changes may confuse the (somewhat brittle) analysis that allow the length of `_v3storage` to sometimes become known in the optimizer.

> * Simplifies `length2`, `dot`, `isInfinite` and `isNaN`.
> * Optimizes `reflect`, `relativeError` and `absoluteError` so that they don't create extra `Vector3` objects.
> * Orders all operations made on `_v3storage` from 1 to 0 so that it is consistent across the class and guarantees that bounds checking is only done once.

These all seem like nice changes.

**Comment by johnmccutchan** [comment]
While I agree with the spirit of your suggestion @rakudrama I feel like this is likely to break a number of users of this library? How should we proceed?

**Comment by johnmccutchan** [comment]
Actually, let's have that discussion in a separate issue: https://github.com/google/vector_math.dart/issues/298

**Comment by johnmccutchan** [comment]
@rakudrama what changes in this PR do you consider blocking? I'd like to split them off and land the non-controversial ones.

**Comment by rakudrama** [comment]
> @rakudrama what changes in this PR do you consider blocking? I'd like to split them off and land the non-controversial ones.

I'm ok with everything except changes to the constructors (and anything that changes a public interface but as far as I remember, we fixed those).
We can make progress by landing the non-constructor changes.

I am looking at the benchmark that was written to support changing the constructors.
Unfortunately, the compilers optimize a lot of the benchmark away, so it does not really make the argument.
It is instead a test of how good the compiler is as removing unused computations.
I will make comments there to make the benchmark better.

I understand that this CL is just following what was done and approved for Vector2.
It is difficult to keep the library cohesive without a design process to tie together multiple CLs.

**Review by johnmccutchan — APPROVED** [review]

**Pull request merged**